### ISSUE TITLE
[Bug] DD-PPO Batch size possibly out of sync between workers

### ISSUE THREAD

**Issue by dsctt** [ray-project/ray#22796]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

DD-PPO throws `Batch size possibly out of sync between workers` as seen here (https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ddppo.py#L244) whenever `count_steps_by=agent_steps`.

### Versions / Dependencies

python=3.9.7
ray=1.10.0 and copy the ddppo.py file from master (need changes from #22521)
nmmo[rllib]=1.5.3.17
OS=Ubuntu 20.04.3 LTS

### Reproduction script

import os

import nmmo
import numpy as np

import ray
from ray import rllib, tune
from ray.rllib.models.torch.recurrent_net import RecurrentNetwork

import torch
from torch import nn
from torch.nn.utils import rnn

class SmallExploreConfig(nmmo.config.Small, nmmo.config.Resource):
   '''Config for NMMO default environment with concurrent spawns'''
   NUM_WORKERS             = 2
   NUM_GPUS                = 0
   ROLLOUT_FRAGMENT_LENGTH = 128
   TRAIN_HORIZON           = 128
   NPOLICIES               = 1
   SGD_MINIBATCH_SIZE      = 128

   EVALUATE                = False
   N_TRAIN_MAPS            = 256

   #Policy specification
   AGENTS                  = [nmmo.Agent]

   #Hardware and debug
   NUM_GPUS_PER_WORKER     = 0
   LOG_LEVEL               = 1

   #Training and evaluation settings
   TRAINING_ITERATIONS     = 1000
   NUM_SGD_ITER            = 1

   #Reward
   COOPERATIVE             = False
   TEAM_SPIRIT             = 0.0

   @property
   def PATH_MAPS(self):
      maps = super().PATH_MAPS
      if self.EVALUATE:
         self.TERRAIN_FLIP_SEED = True
         return os.path.join(maps, 'evaluation')
      return os.path.join(maps, 'training')

   # Spawn all agents at t=0
   @property
   def SPAWN(self):
      return self.SPAWN_CONCURRENT

class RLlibEnv(nmmo.Env, rllib.MultiAgentEnv):
   '''Wrapper class for using Neural MMO with RLlib'''
   def __init__(self, config):
      self.config = config['config']
      super().__init__(self.config)

   def step(self, decisions):
      obs, rewards, dones, infos = super().step(decisions)
      config = self.config
      ts = config.TEAM_SPIRIT
      
      if config.COOPERATIVE:
          #Union of task rewards across population
          team_rewards = defaultdict(lambda: defaultdict(int))
          populations = {}
          for entID, info in infos.items():
              pop = info.pop('population')
              populations[entID] = pop
              team = team_rewards[pop]
              for task, reward in info.items():
                  team[task] = max(team[task], reward)

          #Team spirit interpolated between agent and team summed task rewards
          for entID, reward in rewards.items():
              pop = populations[entID]
              rewards[entID] = ts*sum(team_rewards[pop].values()) + (1-ts)*reward

      dones['__all__'] = False
      test = config.EVALUATE or config.RENDER
      
      if config.EVALUATE:
         horizon = config.EVALUATION_HORIZON
      else:
         horizon = config.TRAIN_HORIZON

      population  = len(self.realm.players) == 0
      hit_horizon = self.realm.tick >= horizon
      
      if not config.RENDER and (hit_horizon or population):
         dones['__all__'] = True

      return obs, rewards, dones, infos
   
   def reward(self, player):
      # Default survival reward
      reward, info = super().reward(player)

      # Inject exploration attribute into player
      if not hasattr(player, 'exploration'):
         player.exploration = 0

      # Historical exploration already part of player state
      exploration = player.history.exploration

      # Only reward agents for distance increment
      # over previous farthest exploration
      if exploration > player.exploration:
         reward += 0.05 * (exploration - player.exploration)

      return reward, info

class Trainer:
   def __init__(self, config, env=None, logger_creator=None):
      super().__init__(config, env, logger_creator)

   @classmethod
   def name(cls):
      return cls.__bases__[0].__name__

   def post_mean(self, stats):
      for key, vals in stats.items():
          if type(vals) == list:
              stats[key] = np.mean(vals)

   def train(self):
      stats = super().train()
      self.post_mean(stats['custom_metrics'])
      return stats

def PPO(config):
   class PPO(Trainer, rllib.agents.ppo.ppo.PPOTrainer): pass
   extra_config = {'sgd_minibatch_size': config.SGD_MINIBATCH_SIZE}
   algo_config = {'TRAIN_BATCH_SIZE': 128}
   return PPO, extra_config, algo_config

def DDPPO(config):
   class DDPPO(Trainer, rllib.agents.ppo.ddppo.DDPPOTrainer): pass
   algo_config = {'TRAIN_BATCH_SIZE': -1}
   return DDPPO, {}, algo_config

def run_tune_experiment(config, trainer_wrapper, rllib_env):

   #Register custom env and policies
   ray.tune.registry.register_env("Neural_MMO",
         lambda config: rllib_env(config))

   mapPolicy = lambda agentID : 'policy_{}'.format(
         agentID % config.NPOLICIES)

   policies = {}
   env = nmmo.Env(config)
   for i in range(config.NPOLICIES):
      params = {
            "agent_id": i,
            "obs_space_dict": env.observation_space(i),
            "act_space_dict": env.action_space(i)}
      key           = mapPolicy(i)
      policies[key] = (None, env.observation_space(i), env.action_space(i), params)

   trainer_cls, extra_config, algo_config = trainer_wrapper(config)

   #Create rllib config
   rllib_config = {
      'num_workers': config.NUM_WORKERS,
      'num_gpus_per_worker': config.NUM_GPUS_PER_WORKER,
      'num_gpus': config.NUM_GPUS,
      'num_envs_per_worker': 1,
      'train_batch_size': algo_config['TRAIN_BATCH_SIZE'],
      'rollout_fragment_length': config.ROLLOUT_FRAGMENT_LENGTH,
      'num_sgd_iter': config.NUM_SGD_ITER,
      'framework': 'torch',
      'horizon': np.inf,
      'soft_horizon': False, 
      'no_done_at_end': False,
      'env': 'Neural_MMO',
      'env_config': {
         'config': config
      },
      'multiagent': {
         'policies': policies,
         'policy_mapping_fn': mapPolicy,
         'count_steps_by': 'agent_steps'
      },
      "model": {
        "fcnet_hiddens": [64, 64],
        "fcnet_activation": "relu",
      },
      'render_env': False,
   }

   #Alg-specific params
   rllib_config = {**rllib_config, **extra_config}

   tune.run(trainer_cls,
      config    = rllib_config,
      name      = trainer_cls.name(),
      verbose   = config.LOG_LEVEL,
      stop      = {'training_iteration': config.TRAINING_ITERATIONS},
      reuse_actors = True,
      )

if __name__ == '__main__':
    run_tune_experiment(SmallExploreConfig(), DDPPO, RLlibEnv)

### Anything else

Some potentially useful context:

I noticed that the expected batch size was consistently much bigger than what was received. I believe this is due to `batch.count` counting env_steps instead of agent_steps as seen here (https://github.com/ray-project/ray/blob/master/rllib/evaluation/rollout_worker.py#L836).

When I adjust the DD-PPO code to take this into account, the received size gets closer to what is expected but is still off by some seemingly random amount. So I tried just removing the assertion altogether after not finding anything similar in a cursory review of the PPO code (and noticing this same unexpected random delta when running PPO with agent_steps). After removing it, DD-PPO seems to run more like expected (although still can't make sense of the random delta in either PPO or DD-PPO). (Note: the attached script is the smallest that can illustrate the error but it is expected that it doesn't do much after that.)

Hopefully this helps illuminate the issue. As far as I can tell, this appears to be a bug but please let me know if it turns out there is something I have overlooked.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

**Comment by gjoliver** [comment]
it feels like we should just not have this check since by nature dd workers can have different rollouts.
